r-data-manipulation.Rmd

```diff
@@ -0,0 +1,1111 @@
+
+Joins are a way to combine two datasets together by using a column the datasets have in common - like the columns that contain the NTA codes.  Let's work with some smaller datasets to see how this works.  
+```{r, include = FALSE}
+dogs <- tribble(
+  ~animal_name,  ~nta, ~breed_name,
+  "PRECIOUS",  "QN12",  "Yorkshire Terrier",
+  "TOBY",      "QN55",  "Jack Russell Terrier",
+  "GUCCI",     "QN01",  "Yorkshire Terrier",
+  "HUNTER",    "QN12",  "Unknown", 
+)
+neighborhoods <- nta_pop %>% 
+  filter(nta_code %in% c("QN01", "QN02", "QN03", "QN12"), year == 2010) %>% 
+  select(nta = nta_code, population, nta_name)
+```
+
+One dataset might have rows that correspond to dogs, and their neighborhood:
+```{r}
+dogs
+```
+
+Another dataset might have rows that correspond to neighborhoods and their populations:
+```{r}
+neighborhoods 
+```
+Notice the datasets both have a column called `nta`, and this column contains the NTA code. {TODO: either point out this isn't true in full size data, or make it true by changing data.} This provides a way to match rows in `dogs` to the rows in `neighborhoods`: for each row of `dogs`, look up the value of `nta` in `neighborhoods` and return the values of `population` and `nta_name`.  This might result in a new dataset that looks like this: 
+```{r, echo = FALSE, message = FALSE}
+dogs %>% left_join(neighborhoods)
+```
+Both "PRECIOUS" and "HUNTER" live in Hammels-Arverne-Edgemere with a population of 36,885 and "GUCCI" lives in South Jamaica with a population of 38,894.  Notice that "TOBY" has missing values for both `population` and `nta_name` because his value of `nta`, `"QN55"`, wasn't in this small `neighborhoods` data.  
+
+This join can be performed with the `left_join()` function in dplyr:
+```{r}
+dogs %>% left_join(neighborhoods)
+```
+It's called a `left_join()` because we will get (at least) one row back for every row in the dataset on the left, in this case, `dogs`.  I like to think about `left_join()` working just as we described above: for each row in the dataset on the left, look up the value in the dataset on the right. 
+
+Notice that `left_join()` gave us the message `Joining, by = "nta"`.  The join functions in dplyr will use any columns with the same name in both datasets to do the join.  If the datasets have columns for the same quantity but with different names, for instance, if `neighborhoods` had a column called `nta_code` instead of `nta`:
+```{r}
+neighborhoods_2 <- neighborhoods %>% 
+  rename(nta_code = nta)
+neighborhoods_2 
+```
+
+Then attempting the join will result in an error - `data sources have no common variables`:
+```{r, error = TRUE}
+dogs %>% left_join(neighborhoods_2)
+```
+You can use the `by` argument to the join function to specify which columns are to be used to do the join:
+```{r}
+dogs %>% 
+  left_join(neighborhoods_2, by = c("nta" = "nta_code"))
```

Comment:
Might need to explain what the code for the `by` argument is doing and how to use.

Reply:
Joins removed for now. Conversation moved to #186